### DISCUSSION TITLE
Action showing permission denied

### DISCUSSION THREAD

**Discussion by aditya-mitra** [community/community#26239] [Actions]
<p>I had many environmental variables so I used to gpg to encrypt those variables in a file.</p>
<p>I then made a script to decrypt that file and to store it the <em>backend</em> folder of my project during <strong>the run of the action</strong></p>
<p><em>All this are to happen in the github action (see below the yml file)</em></p>
<p>That script of mine looks something like this:</p>
<pre><code class="lang-shell">#!/bin/sh

gpg --quiet --batch --yes --decrypt --passphrase="$DECRYPTER_PASSPHRASE" \
--output ./backend/.env ./.github/secrets/backend.env.gpg
</code></pre>
<p>My script is <strong>running well</strong> when I am executing this command in the project root</p>
<pre><code class="lang-auto">./.github/scripts/backend_decrypt.sh
</code></pre>
<p><strong>But it fails during the github action</strong></p>
<p>My <code>backend-deploy.yml</code> file is like this:</p>
<pre><code class="lang-auto">
name: Backend Deploy

on:
  push:
    branches:
      - master
    paths:
      - 'backend/**'

jobs:
  backendDeploy:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - shell: bash
        env:
          DECRYPTER_PASSPHRASE: ${{secrets.DECRYPTER_PASSPHRASE}}
        run: &gt;
          ./.github/scripts/backend_decrypt.sh
# fails at this point
      - run: |
          cat backend/.env
          pwd
          ls -a
          ls backend/
      - run: &gt;
          ls -a
          ls -a backend/
          echo "Deployed to code-rush-backend.herokuapp.com"
</code></pre>
<p>It is giving me this error:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/c/ca4132c7e1f36a4e5b7c0898eb65442a5a97458a.png" data-download-href="/uploads/short-url/sRe8qjxWm5sAs01VZ7CN4tD9UHU.png?dl=1" title="showing error -&amp;gt; /home/runner/work/_temp/27cb9d1d-f99d-489b-b13e-4f56e86348af.sh: line 1: ./.github/scripts/backend_decrypt.sh: Permission denied" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/55396651/181096368-50189809-b56d-4286-b405-c4990bd9a51b.png" alt="showing error -&gt; /home/runner/work/_temp/27cb9d1d-f99d-489b-b13e-4f56e86348af.sh: line 1: ./.github/scripts/backend_decrypt.sh: Permission denied" data-base62-sha1="sRe8qjxWm5sAs01VZ7CN4tD9UHU" width="690" height="160" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/c/ca4132c7e1f36a4e5b7c0898eb65442a5a97458a_2_690x160.png, https://aws1.discourse-cdn.com/github/optimized/2X/c/ca4132c7e1f36a4e5b7c0898eb65442a5a97458a_2_1035x240.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/c/ca4132c7e1f36a4e5b7c0898eb65442a5a97458a.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/c/ca4132c7e1f36a4e5b7c0898eb65442a5a97458a_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">showing error -&amp;gt; /home/runner/work/_temp/27cb9d1d-f99d-489b-b13e-4f56e86348af.sh: line 1: ./.github/scripts/backend_decrypt.sh: Permission denied</span><span class="informations">1362×317 19.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<h2>How can I fix this ? Please help</h2>

**Comment by airtower-luna** [comment, marked as answer]
<p>That most likely means that your <code>./.github/scripts/backend_decrypt.sh</code> script doesn’t have the “execute” filesystem permission set. I assume you’re developing on Windows locally, which doesn’t have that kind of permission system.</p>
<p>You can tell git to add the permission anyway with this command:</p>
<pre><code class="lang-bash">git update-index --chmod=+x ./.github/scripts/backend_decrypt.sh
</code></pre>
<p><code>update-index</code> is similar to <code>add</code> in that it adds the change to the index, so you’ll have to commit and push as usual.</p>

**Reply by issqTK** [reply]
Hi, 
I am having an issue here, after tapping this command on my terminal, I got this **Error**:
_error: .github/scripts/backend_decrypt.sh: does not exist and --remove not passed
fatal: Unable to process path .github/scripts/backend_decrypt.sh_

**Reply by airtower-luna** [reply]
Is `.github/scripts/backend_decrypt.sh` actually the file you need to mark as executable?

**Reply by caldempsey** [reply]
i ❤️  u

**Reply by AliArmanLMG** [reply]
you are a life saver

**Reply by rjparton** [reply]
Yes honestly thank you!

**Comment by aditya-mitra** [comment]
<p>Thanks a lot <a class="mention" href="/airtower-luna">@airtower-luna</a><br>
This solved it</p>

**Comment by manpreetsinghminhas** [comment]
<p>Thank you so much! This solved the issue for me too! Any tips on how to do this by default on a Windows machine for bash scripts?</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="manpreetsinghminhas" data-post="4" data-topic="134957">
<div class="title">
<div class="quote-controls"></div>
 manpreetsinghminhas:</div>
<blockquote>
<p>Any tips on how to do this by default on a Windows machine for bash scripts?</p>
</blockquote>
</aside>
<p>How do you mean “by default”? If you mean marking scripts as executable, no. Unix-ish systems don’t do that either, it’s just that there you automatically notice if something isn’t marked as executable but should. You could try to make a pre-commit hook with some sort of heuristic to detect scripts, but I doubt it’s worth the effort.</p>

**Comment by glenntroncquo** [comment]
<p>Thanks <a class="mention" href="/airtower-luna">@airtower-luna</a><br>
That fixed it for me</p>

**Comment by King-Dorji** [comment]
<p>This one worked very well for me too, thanks</p>

**Comment by ngohuytrieu** [comment]
Hi guys, how about ubuntu?
I am trying to ssh to AWS EC2 but my `deploy.sh is permission dennied`
```
deploy:
    runs-on: ubuntu-18.04
    steps:
      - uses: actions/checkout@v2
      - name: SSH to AWS EC2
        uses: shimataro/ssh-key-action@v2
        with:
          key: ${{ secrets.SSH_KEY }}
          name: id_rsa
          known_hosts: ${{ secrets.SSH_HOSTNAME }}
      - name: Deploy
        run: ./deploy.sh
```

**Reply by airtower-luna** [reply]
Depends on the exact output that'd tell you *what* was denied. If you need help posting a link is best, copy the step output if the repo is private (censor anything private). If the problem is that `deploy.sh` isn't marked executable my solution should work. :slightly_smiling_face: 

If it's anything else… That might warrant a separate discussion. :wink:

**Reply by ngohuytrieu** [reply]
@airtower-luna I've got that fixed, but now I get another issue
```
Run ssh -i ~/.ssh/private.key -tt username@*** "
Host key verification failed.
Error: Process completed with exit code 255.
```
Here is my actions:
```
- name: Create SSH key
        run: |
          mkdir -p ~/.ssh/
          echo "$SSH_PRIVATE_KEY" > ~/.ssh/private.key
          sudo chmod 600 ~/.ssh/private.key
          echo "$SSH_KNOWN_HOSTS" > ~/.ssh/known_hosts
          sudo chmod 600 ~/.ssh/known_hosts
        shell: bash
        env:
          SSH_PRIVATE_KEY: ${{secrets.SSH_KEY}}
          SSH_KNOWN_HOSTS: ${{secrets.SSH_HOSTNAME}}
      - name: SSH into EC2
        run: >
          ssh -i ~/.ssh/private.key -tt username@hostname
          "
            echo 'Welcome to EC2'
          "
```

**Reply by airtower-luna** [reply]
> Host key verification failed.

This means that the key with which the host identifies itself couldn't be recognized. That usually means one of two things:

1. Something (or someone) is messing with your connection.
2. Your `~/.ssh/known_hosts` is incorrect.

This part of your file makes me suspect the latter:
```
          SSH_KNOWN_HOSTS: ${{secrets.SSH_HOSTNAME}}
```
The known hosts file needs to contain not only the hostname, but also the associated public key information, see: https://manpages.debian.org/bullseye/openssh-server/sshd.8.en.html#SSH_KNOWN_HOSTS_FILE_FORMAT

As I said, please create *a new discussion* if you need more help, because this is a completely different topic from the original post. I will not be posting further off-topic responses here.

**Reply by renshengyiyijihe** [reply]
on Linux or macOS, you can run the following command to set execution permission on the file:
`chmod +x script.sh`
